src/AsyncEventSource.h

```diff
@@ -53,58 +60,150 @@ class AsyncEventSourceResponse;
+     *
+     * @param client
+     * @return size_t returns num of bytes the clien was able to send()
+     */
     size_t send(AsyncClient* client);
-    bool finished() { return _acked == _len; }
-    bool sent() { return _sent == _len; }
+
+    // returns true if full message's length were acked
+    bool finished() { return _acked == _data->length(); }
+
+    /**
+     * @brief returns true if all data has been sent already
+     *
+     */
+    bool sent() { return _sent == _data->length(); }
 };
 
+/**
+ * @brief class holds a sse messages queue for a particular client's connection
+ *
+ */
 class AsyncEventSourceClient {
   private:
     AsyncClient* _client;
     AsyncEventSource* _server;
-    uint32_t _lastId;
+    uint32_t _lastId{0};
+    size_t _inflight{0};                   // num of unacknowledged bytes that has been written to socket buffer
+    size_t _max_inflight{SSE_MAX_INFLIGH}; // max num of unacknowledged bytes that could be written to socket buffer
     std::list<AsyncEventSourceMessage> _messageQueue;
 #ifdef ESP32
     mutable std::mutex _lockmq;
 #endif
     bool _queueMessage(const char* message, size_t len);
+    bool _queueMessage(AsyncEvent_SharedData_t&& msg);
     void _runQueue();
 
   public:
     AsyncEventSourceClient(AsyncWebServerRequest* request, AsyncEventSource* server);
     ~AsyncEventSourceClient();
 
-    AsyncClient* client() { return _client; }
-    void close();
-    bool write(const char* message, size_t len);
+    /**
+     * @brief Send an SSE message to client
+     * it will craft an SSE message and place it to client's message queue
+     *
+     * @param message body string, could be single or multi-line string sepprated by \n, \r, \r\n
+     * @param event body string, a sinle line string
+     * @param id sequence id
+     * @param reconnect client's reconnect timeout
+     * @return true if message was placed in a queue
+     * @return false if queue is full
+     */
+    bool send(const char* message, const char* event = NULL, uint32_t id = 0, uint32_t reconnect = 0);
     bool send(const String& message, const String& event, uint32_t id = 0, uint32_t reconnect = 0) { return send(message.c_str(), event.c_str(), id, reconnect); }
     bool send(const String& message, const char* event, uint32_t id = 0, uint32_t reconnect = 0) { return send(message.c_str(), event, id, reconnect); }
-    bool send(const char* message, const char* event = NULL, uint32_t id = 0, uint32_t reconnect = 0);
+
+    /**
+     * @brief place supplied preformatted SSE message to the message queue
+     * @note message must a properly formatted SSE string according to https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events
+     *
+     * @param message data
+     * @return true on success
+     * @return false on queue overflow or no client connected
+     */
+    bool write(AsyncEvent_SharedData_t message) { return connected() && _queueMessage(std::move(message)); };
```

Comment:
same here: it leaks to the public api `std::string`...

Reply:
it must be there for `AsyncEventSource` class to enqueue a shallow copy (a shared pointer) message to all existing queues in case if there are multiple clients are connected.
If enqueued via `const char*`  it would end up with deep-copy clones in each queue.